File: src/execution/redirections.c
#include "tokens.h"
#include <fcntl.h>
#include <unistd.h>
#include <errno.h>
#include <unistd.h>

static int	open_redir(const char *file, enum e_redir redir, int *fd)
{
	int		perm;

	if (redir == INPUT_REDIR)
	{
		perm = access(file, F_OK);
		if (perm != 0)
			return (127);
		perm = access(file, R_OK);
		if (perm != 0)
			return (126);
		*fd = open(file, O_RDONLY);
	}
	else if (redir == OUTPUT_REDIR)
		*fd = open(file, O_WRONLY | O_CREAT | O_TRUNC, 0644);
	else if (redir == APPEND)
	{
		perm = access(file, W_OK);
		if (perm != 0)
			return (errno);
		*fd = open(file, O_WRONLY | O_CREAT | O_APPEND, 0644);
	}
	if (*fd == -1)
		return (errno);
	return (0);
}

/**
 * @brief
 *
 * @param token
 * @return int 0 if no error, 1 if open failed, 2 if access failed
 */
int	do_redirections(t_arg *cmd_args, char **error_elem)
{
	int	i;
	int	fd;

	i = 0;
	while (cmd_args[i].elem)
	{
		fd = 0;
		if (cmd_args[i].type == REDIR && cmd_args[i].redir != HEREDOC)
		{
			if (open_redir(cmd_args[i].elem, cmd_args[i].redir, &fd) != 0)
			{
				*error_elem = cmd_args[i].elem;
				return (errno);
			}
			if (cmd_args[i].redir != INPUT_REDIR)
				dup2(fd, STDOUT_FILENO);
			else
				dup2(fd, STDIN_FILENO);
			close(fd);
		}
		i++;
	}
	return (0);
}

File: src/execution/redirections.rs
use crate::msh::{
	ArgType::Redir,
	CommandArg,
	RedirType::{Append, HereDoc, InputRedir, OutputRedir},
	eprint_msh,
};
use nix::{
	fcntl::{OFlag, open},
	sys::stat::Mode,
	unistd::{AccessFlags, access},
};
use std::ffi::CString;

pub(super) fn do_redirections(cmd_args: &mut [CommandArg]) -> Result<(), i32> {
	let mut i = 0;
	while i < cmd_args.len() {
		if (cmd_args[i]).type_0 == Some(Redir) && (cmd_args[i]).redir.unwrap() != HereDoc {
			let file = CString::new((cmd_args[i]).elem_str.clone()).unwrap();
			let file = file.as_c_str();
			let fd_result = match (cmd_args[i]).redir.unwrap() {
				InputRedir => {
					if access(file, AccessFlags::F_OK).is_err() {
						return Err(127);
					} else if access(file, AccessFlags::R_OK).is_err() {
						return Err(126);
					}
					open(file, OFlag::O_RDONLY, Mode::empty())
				}
				OutputRedir => open(
					file,
					OFlag::O_WRONLY | OFlag::O_CREAT | OFlag::O_TRUNC,
					Mode::from_bits(0o644).unwrap(),
				),
				Append => {
					if access(file, AccessFlags::W_OK).is_err() {
						return Err(126);
					}
					open(
						file,
						OFlag::O_WRONLY | OFlag::O_CREAT | OFlag::O_APPEND,
						Mode::from_bits(0o644).unwrap(),
					)
				}
				HereDoc => unreachable!(),
			};
			if let Ok(fd) = fd_result {
				if (cmd_args[i]).redir == Some(InputRedir) {
					let _ = nix::unistd::dup2_stdin(&fd);
				} else {
					let _ = nix::unistd::dup2_stdout(&fd);
				}
				let _ = nix::unistd::close(fd);
			} else {
				eprint_msh!("failed to execute: {}", (cmd_args[i]).elem_str);
				return Err(-1);
			}
		}
		i += 1;
	}
	Ok(())
}
